Upgrade Azure repository SDK to v12

fcofdez opened the pull request:
This PR upgrades the Azure repository SDK to v12. 

I've added comments where I thought we could discuss if a different approach would make sense or some parts that might need more context. 

Additionally I've opened a few issues on the azure SDK side to overcome some problems that I've encountered along the way:
- https://github.com/Azure/azure-sdk-for-java/issues/1727
- https://github.com/Azure/azure-sdk-for-java/issues/17590
- https://github.com/Azure/azure-sdk-for-java/pull/17627

Sorry for the size of this PR. There's quite a bit of boilerplate regarding new licenses and so on and I couldn't think of a way to implement this in smaller chunks.

fcofdez committed:
Upgrade azure sdk

fcofdez committed:
Fix tests

fcofdez committed:
Delegate to regular scheduler instead of into a custom one

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/executors/ReactorScheduledExecutorService.java:
```diff
@@ -0,0 +1,188 @@
+
+        return threadPool.scheduler().scheduleAtFixedRate(() -> {
+            try {
+                delegate.execute(decoratedCommand);
```
It's possible that the command is executed some time after the specified `period` if the thread pool backed by `delegate` is over subscribed. This is used for timeouts and retries, I guess we could live with a less precise timers but still control the executor in which that code is executed.

fcofdez commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java:
```diff
@@ -215,11 +234,49 @@ public void testReadBlobWithRetries() throws Exception {
+        }
+    }
+
     public void testReadRangeBlobWithRetries() throws Exception {
```
This test fails right now, since the SDK triggers a `HEAD` request to get the blob size. The problem comes when a range that goes beyond the blob end the API would return an error `416 (Requested Range Not Satisfiable)`. My guess is that the expectation is that client would know the blob length beforehand and won't request past that boundary, is that correct?

fcofdez replied:
Additionally, we could be a little bit more efficient and avoid a copy + allocation per chunk read request if we implement it ourselves.

original-brownbear replied:
As discussed, let's do it and try to avoid the extra HEAD + copying if we can.

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureRepositoryPlugin.java:
```diff
@@ -100,6 +150,32 @@ public void reload(Settings settings) {
+    }
+
+    @SuppressForbidden(reason = "needs to overcome the eager class loading problem")
+    private void setUpDefaultHttpClientProvider() {
```
This is really unfortunate, the SDK loads a default http client provider during the class construction and since the dependency jars are loaded in lexicographical order, the netty http client is not loaded at that time. This is a bug on the blob batch client, that instead of reusing the existing http client, it creates a new one. I've opened a PR and it's already approved, so this is just a temporal workaround.

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:
```diff
@@ -0,0 +1,295 @@
+        // Most of the code that needs special permissions (i.e. jackson serializers generation) is executed
+        // in the event loop executor. That's the reason why we should provide an executor that allows the
+        // execution of privileged code
+        final EventLoopGroup eventLoopGroup = new NioEventLoopGroup(eventLoopThreadsFromSettings(settings),
```
As I point out here most of the response handling code is executed in the netty event loop executor, meaning that we should provide a privileged environment to the tasks executed in that event loop. 
I tried different approaches to solve this issue, one of them was to create a wrapper around 
 `com.azure.core.http.HttpClient` that would publish the emitted response in a different thread pool:

```java
Mono<HttpResponse> send(HttpRequest request) {
         return delegate.send(request).publishOn(privilegedExecutor);
}
```

But it seems like the `HttpResponse` is parsed before, so that didn't help. 

Maybe I'm missing something here and we can provide a more fine-grained solution.

fcofdez commented on test/fixtures/azure-fixture/src/main/java/fixture/azure/AzureHttpHandler.java:
```diff
@@ -50,11 +54,17 @@
     private final String container;
 
-    public AzureHttpHandler(final String container) {
+    public AzureHttpHandler(final String account, final String container) {
```
In te latest SDK if the endpoint uses an ip host, it expects the account in the URI path, see `com.azure.storage.blob.BlobUrlParts#parseIpUrl` (it doesn't validate it there, but it fails later on as it expects the blob name to be on the 3 part of the path, i.e. `http://192.168.1.20:8080/account/container/blob`.)

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:
```diff
@@ -0,0 +1,295 @@
+
+    private final ThreadPool threadPool;
+    private final String reactorExecutorName;
+    private final EventLoopGroup eventLoopGroup;
```
All this resources are shared by all the clients, so we only need to manage the lifecycle of this component and the clients can be stateless.

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:
```diff
@@ -0,0 +1,295 @@
+    private static final int DEFAULT_MAX_CONNECTIONS = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
+    private static final int DEFAULT_EVENT_LOOP_THREAD_COUNT = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
+
+    static final Setting<String> EVENT_LOOP_EXECUTOR = Setting.simpleString(
```
Is this a good way to define those settings? I think these are mostly node related so it makes sense to me to configure them at that level.

original-brownbear replied:
@fcofdez ping :) this is still an open question I have.

fcofdez replied:
That's to simplify testing, so we can pass a different executor there, in scenarios where we don't create the ones from the Azure plugin.

fcofdez replied:
And sorry, I missed this comment

fcofdez replied:
Addressed in https://github.com/elastic/elasticsearch/pull/65140/commits/2c19d717a34a410b5f749bf23d0770a01e6e7b2a

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:
```diff
@@ -0,0 +1,295 @@
+    private static final TimeValue DEFAULT_CONNECTION_TIMEOUT = TimeValue.timeValueSeconds(30);
+    private static final TimeValue DEFAULT_MAX_CONNECTION_IDLE_TIME = TimeValue.timeValueSeconds(60);
+    private static final int DEFAULT_MAX_CONNECTIONS = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
+    private static final int DEFAULT_EVENT_LOOP_THREAD_COUNT = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
```
Maybe this is too much? It's capped to 16

original-brownbear replied:
Small aside, can go into a follow-up (I can deal with it later this week): I think we can make this `1` for now. The consumers that consume whatever buffers this pool produces will never be able to grind through what a single network thread can produce in throughput and the memory savings from reducing the thread count are non-trivial here. We can do a few official benchmarks to this effect though :)

fcofdez committed:
Update default max connections.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

fcofdez committed:
More efficient blob reading

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Take into account retries during reads

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Fix failing test

fcofdez commented:
jenkins test this

original-brownbear commented:
@fcofdez there seems to be some bug with deleting blobs left here?

CI fails on some of the delete blob tests and this fails for me locally as well:

```
/gradlew 'azure-test-goal-that-idea-doesnt-shotw' --tests "org.elasticsearch.repositories.azure.AzureBlobStoreRepositoryTests.testDeleteBlobs" -Dtests.seed=D55C7FBE34D4B07E -Dtests.locale=ru -Dtests.timezone=WET -Druntime.java=15
```

fcofdez commented:
Yes, it's related with a change introduced recently in the security side that restricts the permissions that can be granted to plugins. https://github.com/elastic/elasticsearch/pull/64751. I'm waiting to see if there's a workaround for this.

fcofdez committed:
Grant permissions to tasks executed in reactor schedulers.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez commented:
Sorry @original-brownbear, you were right. I thought I was providing the proper permissions to all tasks, but I missed the reactor schedulers. It should be fixed now.

original-brownbear commented:
No worries @fcofdez !

Unfortunately it looks like things are still broken due to permissions trouble, see failing test:

```
13:20:11 > Task :x-pack:plugin:searchable-snapshots:qa:azure:integTest FAILED
13:20:11 Exec output and error:
13:20:11 | Output for ./bin/elasticsearch-plugin:-> Installing file:/dev/shm/elastic+elasticsearch+pull-request-2/plugins/repository-azure/build/distributions/repository-azure-8.0.0-SNAPSHOT.zip
13:20:11 | -> Downloading file:/dev/shm/elastic elasticsearch pull-request-2/plugins/repository-azure/build/distributions/repository-azure-8.0.0-SNAPSHOT.zip
13:20:11 | -> Failed installing file:/dev/shm/elastic+elasticsearch+pull-request-2/plugins/repository-azure/build/distributions/repository-azure-8.0.0-SNAPSHOT.zip
13:20:11 | -> Rolling back file:/dev/shm/elastic+elasticsearch+pull-request-2/plugins/repository-azure/build/distributions/repository-azure-8.0.0-SNAPSHOT.zip
13:20:11 | -> Rolled back file:/dev/shm/elastic+elasticsearch+pull-request-2/plugins/repository-azure/build/distributions/repository-azure-8.0.0-SNAPSHOT.zip
13:20:11 | Exception in thread "main" java.lang.IllegalArgumentException: plugin policy [/dev/shm/elastic+elasticsearch+pull-request-2/x-pack/plugin/searchable-snapshots/qa/azure/build/testclusters/integTest-0/distro/8.0.0-DEFAULT/plugins/.installing-15635625663194161767/plugin-security.policy] contains illegal permission ("java.lang.reflect.ReflectPermission" "newProxyInPackage.com.azure.storage.blob.implementation") in global grant
13:20:11 | 	at org.elasticsearch.bootstrap.PolicyUtil.validatePolicyPermissionsForJar(PolicyUtil.java:311)
13:20:11 | 	at org.elasticsearch.bootstrap.PolicyUtil.validatePolicyPermissions(PolicyUtil.java:321)
13:20:11 | 	at org.elasticsearch.bootstrap.PolicyUtil.getPluginPolicyInfo(PolicyUtil.java:332)
13:20:11 | 	at org.elasticsearch.plugins.InstallPluginCommand.installPlugin(InstallPluginCommand.java:854)
13:20:11 | 	at org.elasticsearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:240)
13:20:11 | 	at org.elasticsearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:210)
13:20:11 | 	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
13:20:11 | 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:127)
13:20:11 | 	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:91)
13:20:11 | 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:127)
13:20:11 | 	at org.elasticsearch.cli.Command.main(Command.java:90)
13:20:11 | 	at org.elasticsearch.plugins.PluginCli.main(PluginCli.java:47)
13:20:11 
13:20:11 > Task :x-pack:plugin:ml:thirdPartyAudit
````

original-brownbear reviewed (commented):
Thanks Francisco: a few smaller comments. Sorry for being so slow here, I'll pick this up at a faster pace tomorrow!

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:
```diff
@@ -153,213 +200,263 @@ public LocationMode getLocationMode() {
+
+                List<Mono<Response<Void>>> batchResponses = new ArrayList<>(batches.size());
+                for (BlobBatch batch : batches) {
+                    batchResponses.add(blobBatchClient.submitBatchWithResponse(batch, false));
```
You're right, I've just tested this using a SAS token and I got a:
```
com.azure.storage.blob.models.BlobStorageException: Status code 403, "<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthorizationResourceTypeMismatch</Code><Message>This request is not authorized to perform this operation using this resource type.
```

I'll implement a different path that triggers all the delete requests when a SAS token is used :(.

original-brownbear replied:
> I'll implement a different path that triggers all the delete requests when a SAS token is used :(.

Let's just use a single path IMO. No need to optimize for non-token credentials, those are pretty rare in production.
I think we can still get into a much better state than what we had with the v8 SDK because we can just fire off all the deletes at once with the async API?

fcofdez replied:
👍, yes that was my plan.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Use regular delete requests instead of the batch API

fcofdez committed:
Initialize jackson parser during AzureRepositoryPlugin class
initialization

fcofdez commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureRepositoryPlugin.java:
```diff
@@ -86,11 +131,16 @@ AzureStorageService createAzureStoreService(final Settings settings) {
 
     public static ExecutorBuilder<?> executorBuilder() {
-        return new ScalingExecutorBuilder(REPOSITORY_THREAD_POOL_NAME, 0, 32, TimeValue.timeValueSeconds(30L));
+        return new ScalingExecutorBuilder(REPOSITORY_THREAD_POOL_NAME, 0, 5, TimeValue.timeValueSeconds(30L));
```
We use this thread pool to execute some of the reactor tasks, such as timeout or retries handlers (the scheduling is done using `org.elasticsearch.threadpool.ThreadPool#scheduler` but the tasks itself are executed in this thread pool)

rjernst committed:
Force azure impl to be public

The azure sdk internally dynamically constructs its client classes.
However, one of these, for the blob storage is private. This has been
fixed upstream, but until that is released, the client is unuseable
without the newProxyInPackage permission. Rather than grant that
permission, this commit makes the class in question public by patching
the azure class file when building the azure repository plugin.

relates https://github.com/Azure/azure-sdk-for-java/issues/17368

fcofdez committed:
Make public additional sdk interfaces to avoid permission issues

fcofdez committed:
Fix precommit error

rjernst committed:
Fix wiring of tasks

fcofdez committed:
Handle timeout configuration correctly

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Remove unused dependency

fcofdez committed:
Add tests for LocationMode configurations

fcofdez committed:
Make InnerClass marker public too durin the azure sdk services rewrite.

fcofdez committed:
Add tests for using SECONDARY_ONLY endpoints

fcofdez commented:
@rjernst The class rewriting is working now, we were missing marking the containing class `InnerClass` visibility markers as public. It seems like the jvm treats inner classes differently in that regard (https://github.com/openjdk/jdk/blob/3da30e991a741fd985d844afe1be5a09c5d7d4a1/src/hotspot/share/include/jvm.h#L594-L601). Thanks for your help! :)

fcofdez committed:
Add min values to AzureClientProvider settings

rjernst committed:
Cleanup publicifier

rjernst committed:
restore docker maybe...

fcofdez committed:
Read eagerly on AzureInputStream

fcofdez committed:
Add tests for retry strategies

fcofdez committed:
Merge remote-tracking branch 'fcofdez/azure-sdk-upgrade-new' into azure-sdk-upgrade

fcofdez committed:
Fix checkStyle

fcofdez commented:
@elasticmachine please run elasticsearch-ci/bwc

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

original-brownbear commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java:
```diff
@@ -282,7 +365,8 @@ public void testWriteBlobWithRetries() throws Exception {
                         Streams.readFully(exchange.getRequestBody());
-                        AzureHttpHandler.sendError(exchange, randomFrom(RestStatus.INTERNAL_SERVER_ERROR, RestStatus.SERVICE_UNAVAILABLE));
+                        AzureHttpHandler.sendError(exchange, randomFrom(RestStatus.INTERNAL_SERVER_ERROR, RestStatus.SERVICE_UNAVAILABLE
+                        ));
```
NIT: revert line braeak

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:
```diff
@@ -153,213 +194,230 @@ public LocationMode getLocationMode() {
+        } catch (Exception e) {
+            logger.warn("can not access [{}] in container {{}}: {}", blob, container, e.getMessage());
+        }
+        return false;
```
Why do we return `false` if we run into an exception?
This check is used to potentially mark a repository corrupted because an expected `index-N` at the repo root is not found. If we return `false` on some random IO/network exception then we'd potentially mark a repo as corrupted when there's nothing wrong with it?

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:
```diff
@@ -391,96 +449,112 @@ public void eventOccurred(RequestCompletedEvent eventArg) {
+                                 int maxRetries) throws IOException {
+            super(rangeOffset, rangeLength, chunkSize, contentLength);
+            this.client = client;
+            this.buffer = ByteBuffer.allocate(chunkSize);
```
We can size this the same way we size `firstReadLength` so that we don't allocate 4M for a file smaller than 4M right?

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:
```diff
@@ -391,96 +449,112 @@ public void eventOccurred(RequestCompletedEvent eventArg) {
+            // Since we disabled marking in this InputStream, the offset only advances
+            // and never goes back requesting the firstReadOffset again
+            if (offset != firstReadOffset || readLength != firstReadLength) {
+                executeRead(readLength, offset);
```
Sorry for not pointing this out earlier, but I'm a little worried about this stream.
Is it actually faster than what we currently have with the old SDK? We still have the same inefficient behavior here as far as I can tell. We fetch data into a 4M buffer lazily and copy it into the smaller read buffer that comes from other snapshot functionality like we did before, just using non-blocking IO instead of blocking IO in the background.
This is in contrast to the other SDKs where we just execute one request for the full file that we then read in a blocking fashion (thereby avoiding the latency of a request every 4M and getting much better performance).
Can't we build this much more efficiently by starting a `Flux` on the whole stream and then blockingly iterating through it via `toStream()` to get similar behavior to a full blocking read triggered by a single request?

original-brownbear reviewed (commented):
Sorry @fcofdez didn't get through the whole thing today, but I figured I'd add the comment on the stream read asap so we can discuss it. Let me know what you think there.

So far looks great though apart from that, thanks!

fcofdez committed:
Move to read the entire stream instead of reading it in chunks

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez commented:
@original-brownbear as we discussed I've implemented the `AzureInputStream` reading the entire stream instead of requesting it in chunks, the tricky part comes in https://github.com/elastic/elasticsearch/pull/65140/files#diff-5138cdb95334310e4f811afa0bb04f785e7be1280c5812ab6464211274194990 where the interface between the async API and the sync API lies. We still need to copy the buffers, as I didn't find a reasonable way to avoid copying them :(, for that I'm using a buffer pool so we avoid allocating them over and over. Let me know what you think.

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:
```diff
@@ -153,213 +195,233 @@ public LocationMode getLocationMode() {
+            });
+            return Boolean.TRUE.equals(blobExists);
+        } catch (Exception e) {
+            logger.warn("can not access [{}] in container {{}}: {}", blob, container, e.getMessage());
```
You can make this `trace` again, the exception will bubble up and will be logged at warn level upstream anyway

fcofdez replied:
Fixed in https://github.com/elastic/elasticsearch/pull/65140/commits/2c3eefa1117b3371a21f42ea7902058f274fb746

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/CancellableRateLimitedFluxIterator.java:
```diff
@@ -0,0 +1,230 @@
+
+    @Override
+    public T next() {
+        if (hasNext() == false) {
```
Can we document that `hasNext()` is where we block and wait for the IO thread to put something in the queue? This is quite subtle I think :)

fcofdez replied:
Fixed in https://github.com/elastic/elasticsearch/pull/65140/commits/2c3eefa1117b3371a21f42ea7902058f274fb746

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/CancellableRateLimitedFluxIterator.java:
```diff
@@ -0,0 +1,230 @@
+    public boolean hasNext() {
+        for (; ; ) {
+            boolean isDone = done;
+            boolean isQueueEmpty = queue.isEmpty();
```
How is this thread-safe when `ArrayDeque` isn't thread safe? Aren't we putting stuff on this queue on the Netty event-loop threads but running these checks on the `SNAPSHOT` or whatever pool that is reading from the blocking stream?

fcofdez replied:
Since `done` is volatile, in theory the writes must be visible in the `SNAPSHOT` threads. But I agree that this might be tricky, would you prefer to use an `ArrayBlockingQueue`? To be honest I was on the fence with this solution, the `ArrayBlockingQueue` solution would be easier to understand.

fcofdez replied:
I went ahead and replaced the queue in https://github.com/elastic/elasticsearch/pull/65140/commits/2c3eefa1117b3371a21f42ea7902058f274fb746

original-brownbear reviewed (commented):
Hey @fcofdez the idea looks great. I have a question on the implementation around the queue though, I might be misunderstanding something here?

fcofdez committed:
Use a concurrent queue

original-brownbear commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java:
```diff
@@ -19,132 +19,155 @@
      * Maximum allowed blob size in Azure blob store.
      */
-    public static final ByteSizeValue MAX_CHUNK_SIZE = new ByteSizeValue(Constants.MAX_BLOB_SIZE, ByteSizeUnit.BYTES);
+    public static final ByteSizeValue MAX_CHUNK_SIZE = new ByteSizeValue(256, ByteSizeUnit.MB);
```
Only just noticed this: why is this reduced again? We don't want to split into lots of files anymore, is there an issue with the SDK and larger files?

fcofdez replied:
I've addressed it in https://github.com/elastic/elasticsearch/pull/65140/commits/974bc38c0d18ad3321f3abb4652bb098bf84afde

original-brownbear reviewed (commented):
Sorry for the noisy review, but one important comment inline about the max blob size to look at asap IMO if it's not a random oversight

fcofdez committed:
Use the proper max chunk size

fcofdez committed:
Remove unnecessary settings

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Revert line break

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez committed:
Remove dead code

original-brownbear reviewed (approved):
I gave this a bit of a spin locally and it's definitely using a little more memory than I would have hoped (quite a bit of per thread buffer use). I have some ideas to significantly lower this in a follow up (same tried and true fixes we applied on the transport layer already) and I don't think these should be a blocker for now.
-> Assuming CI goes green, LGTM :) Thanks again+so much for grinding through this Francisco, looks great!

fcofdez committed:
Remove unused import

fcofdez committed:
Merge remote-tracking branch 'origin/master' into azure-sdk-upgrade

fcofdez commented:
Thanks for your review and patience @original-brownbear ! :)
